pyiceberg/table/metadata.py

```diff
@@ -230,6 +230,14 @@ def snapshot_by_id(self, snapshot_id: int) -> Optional[Snapshot]:
         """Get the snapshot by snapshot_id."""
         return next((snapshot for snapshot in self.snapshots if snapshot.snapshot_id == snapshot_id), None)
 
+    def latest_snapshot_before_timestamp(self, timestamp_ms: int) -> Optional[Snapshot]:
+        """Get the snapshot right before the given timestamp."""
+        result, prev_timestamp = None, 0
+        for snapshot in self.snapshots:
```

Comment:
Instead of searching for a valid snapshot in self.snapshots (which has all valid snapshots irrespective of branch), I should probably write an API to find all ancestors of current snapshot and iterate over those snapshots only.